### PR TITLE
Add moomin

### PR DESCRIPTION
**Pull request by htpusa**
Following a request to do so, I'm submitting MOOMIN for integration in to the Toolbox.
MOOMIN (Mathematical explOration of Omics data on a MetabolIc Network) is a tool for analysing differential expression data. The paper describing the method:
https://doi.org/10.1093/bioinformatics/btz584

I will look into creating a tutorial for MOOMIN.

**I hereby confirm that I have:**

- [x] Tested my code on my own machine
- [x] Followed the guidelines in the [Contributing Guide](https://opencobra.github.io/cobratoolbox/docs/contributing.html)
- [x] Selected `develop` as a target branch (top left drop-down menu)

*(Note: You may replace [ ] with [X] to check the box)*

### CODE REVIEW

**Comment by rmtfleming** [comment]
Thanks for contributing this code. 

Your test (and possibly others) are failing. You can see why, at least initatilly, by looking at the output from the test. Click on 'Details' beside the test. 
i.e.
https://prince.lcsb.uni.lu/jenkins/job/COBRAToolbox-pr-auto-linux/MATLAB_VER=R2017b,label=prince-slave-linux-01/3061/display/redirect

Look at the console output by clicking on 'Console' in the page linked to above. Here is the direct link:
https://prince.lcsb.uni.lu/jenkins/job/COBRAToolbox-pr-auto-linux/MATLAB_VER=R2018b,label=prince-slave-linux-01/3061/console

At the end of that file I see:

16:44:18 The following tests failed:
16:44:18 testMoomin.m
16:44:18 
16:44:18 The reasons were as follows:
16:44:18 ------------------------------------------------
16:44:18 testMoomin.m:
16:44:18 Error using testMoomin (line 57)
16:44:18 Assertion failed.
16:44:18 ------------------------------------------------

Please check why your test is failing.

Also, it says MOOMIN needs  "a MILP-solver compatible with COBRA", but that currently only IBM CPLEX is supported. Please make the code compatible with other solveCobraMILP compatible solvers. e.g. Get a free academic licence for gurobi.

Contributions to the cobra toolbox should not be tied to only one solver, unless there is a good reason for it.

File: src/dataIntegration/transcriptomics/MOOMIN/moomin.m [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+	enumerate = 1;
+	precision = 7;
+	
+	% suggested solver parameters, override by passing a parameter structure through
+	%	'solverParameters' option
+	tolerance = 1e-6;
+    solverParameters.intTol = tolerance;
+    solverParameters.absMipGapTol = tolerance;
+    solverParameters.feasTol = tolerance;
+    solverParameters.optTol = tolerance;
```
Comment (tpfau) [inline src/dataIntegration/transcriptomics/MOOMIN/moomin.m]: Why don't you use the default COBRA parameters if no explicit parameters are given?
Setting the parameters in this function is (in my opinion) unexpected behavior which can easily lead to misinterpretation of its results.

File: test/verifiedTests/dataIntegration/testMOOMIN/testMoomin.m [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		% there should be only one solution
+		assert(size(MILPsolutions, 1) == 2 && MILPsolutions{2, 1}.stat == 0);
+	
+		% check the MILP
+		A = MILPproblem.A;
+		x = MILPsolutions{1,1}.full;
+		% stoichiometric balance
+		assert(all(abs(A(1:nMets, 1:nRxns) * x(1:nRxns)) < tol));
+		% binary variables worked as intended
+		assert(all((x(1:nRxns) > 0) == (x(nRxns + 1:2 * nRxns))));
```
Comment (tpfau) [inline test/verifiedTests/dataIntegration/testMOOMIN/testMoomin.m]: Does this have to be a strict equality or could solver tolerances allow a deviation?

File: test/verifiedTests/dataIntegration/testMOOMIN/testMoomin.m [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+			'enumerate', 5, 'stoichiometry', 0);
+
+		% there should be only one solution
+		assert(size(MILPsolutions, 1) == 2 && MILPsolutions{2, 1}.stat == 0);
+	
+		% check the MILP
+		A = MILPproblem.A;
+		b = MILPproblem.b;
+		x = MILPsolutions{1,1}.full;
+		assert(all(A(1:nRxns + nMets, :) * x <= b(1:nRxns + nMets)));
```
Comment (tpfau) [inline test/verifiedTests/dataIntegration/testMOOMIN/testMoomin.m]: again, allow for feasibility tolerances.

File: test/verifiedTests/dataIntegration/testMOOMIN/testWriteMoominOutput.m [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+model.frequency = [1; 1; 0.5];
+model.combined = [1; 6; 6];
+expression.GeneID = {'g1'; 'g2'; 'g3'; 'g4'};
+expression.PPDE = [0.9; 0.9; 0.1; 0.1];
+expression.FC = [1; -1; 0.1; -0.1];
+model.expression = expression;
+
+% test with different input variants
+fprintf(' -- Running testWriteMoominOutput ... ');
+writeMoominOutput(model, 'test.out');
```
Comment (tpfau) [inline test/verifiedTests/dataIntegration/testMOOMIN/testWriteMoominOutput.m]: Could you add comparison files to actually check, that the written output is correct?
Like this, the following `writeMoominOutput` file would pass this test:
```
function writeMoominOutput(varargin)
end
```

File: src/dataIntegration/transcriptomics/MOOMIN/moomin.m (hunk 1 above)
Comment (htpusa) [reply src/dataIntegration/transcriptomics/MOOMIN/moomin.m]: I change the parameters to improve performance. The MILP can become quite difficult to solve. But if this is unwanted, I can change it.

File: test/verifiedTests/dataIntegration/testMOOMIN/testMoomin.m (hunk 2 above)
Comment (htpusa) [reply test/verifiedTests/dataIntegration/testMOOMIN/testMoomin.m]: Doesn't need to be, but I was expecting the binary variables to be truly binary.

File: test/verifiedTests/dataIntegration/testMOOMIN/testWriteMoominOutput.m (hunk 4 above)
Comment (htpusa) [reply test/verifiedTests/dataIntegration/testMOOMIN/testWriteMoominOutput.m]: You mean .txt files? Of course, but I'm not sure what would be the proper way to do the comparison?

**Comment by htpusa** [comment]
>Also, it says MOOMIN needs "a MILP-solver compatible with COBRA", but that currently only IBM CPLEX is supported. Please make the code compatible with other solveCobraMILP compatible solvers. e.g. Get a free academic licence for gurobi.

>Contributions to the cobra toolbox should not be tied to only one solver, unless there is a good reason for it.

I tried with Gurobi, and it didn't work. I will have to try to figure out what the problem is but that is going to take some time.

File: test/verifiedTests/dataIntegration/testMOOMIN/testWriteMoominOutput.m (hunk 4 above)
Comment (tpfau) [reply test/verifiedTests/dataIntegration/testMOOMIN/testWriteMoominOutput.m]: Have a look here:
https://github.com/opencobra/cobratoolbox/blob/c72a7bd511f79ebc1b6b1e0421544b22b4a9335d/test/verifiedTests/analysis/testPrint/testPrintRxnFormula.m

File: src/dataIntegration/transcriptomics/MOOMIN/moomin.m (hunk 1 above)
Comment (tpfau) [reply src/dataIntegration/transcriptomics/MOOMIN/moomin.m]: Hmm...
Personally, I'd call this unexpected behaviour.
I would rather put in a "suggested parameters" information in the Documentation and/or a warning when using parameters which deviate from those you suggest. But changing them "silently" in a function (even if only for that function) is kind of odd and can potentially lead to unexpected problems which are then very hard to figure out.

Comment (htpusa) [reply src/dataIntegration/transcriptomics/MOOMIN/moomin.m]: Duly noted, will change.

**Comment by rmtfleming** [comment]
https://prince.lcsb.uni.lu/jenkins/job/COBRAToolbox-pr-auto-windows7/MATLAB_VER=R2016b,label=prince-slave-windows7-01/1512/console

gives the error below, but we can accept the PR now because that test might be failing for other reasons and we would like to clean things up on develop now anyway and we take your word that support for other solvers is added in future.

12:34:37 The following tests failed:
12:34:37 testSWIFTCORE.m
12:34:37 
12:34:37 The reasons were as follows:
12:34:37 ------------------------------------------------
12:34:37 testSWIFTCORE.m:
12:34:37 Error using testSWIFTCORE (line 41)
12:34:37 Assertion failed.
12:34:37 ------------------------------------------------

**Pull request merged**